### PR TITLE
feat: Media container

### PR DESCRIPTION
**Pull request by rubencarvalho**
### Description

This is a modification of our [container](https://cloudscape.design/components/container/?tabId=playground) component, optimised for content-oriented websites and applications.

This change to the container container allows displaying a borderless media element, either on the left or top of the container.

Added a custom hook `useMedia` to update the responsive media properties (like position, width, height, and content). 

### How has this been tested?

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/container/internal.tsx [hunk 1]
```diff
@@ -100,6 +103,10 @@ export default function InternalContainer({
   // In this case we don't want the container to have sticky styles, as only the table header row will show as stuck on scroll.
   const shouldHaveStickyStyles = isSticky && !isMobile;
 
+  function getMediaStyles() {
+    return mediaPosition === 'top' ? { height: mediaHeight } : { width: mediaWidth };
```
Comment (timogasda) [inline src/container/internal.tsx:107]: Why does this need to be a function? We can just say 

```
const mediaStyles = mediaPosition === 'top' ? { height: mediaHeight } : { width: mediaWidth };
```

File: src/container/internal.tsx [hunk 2]
```diff
@@ -109,50 +116,65 @@ export default function InternalContainer({
         styles.root,
         styles[`variant-${variant}`],
         fitHeight && styles['fit-height'],
+        media?.content && (mediaPosition === 'top' ? styles['with-top-media'] : styles['with-side-media']),
```
Comment (timogasda) [inline src/container/internal.tsx:119]: minor: I think I've counted `media?.content` five times here. Might be worth doing `const hasMedia = !!media?.content`.

That also would turn it into a proper boolean, and not `boolean | undefined`

File: src/container/internal.tsx [hunk 3]
```diff
@@ -109,50 +116,65 @@ export default function InternalContainer({
...
+          styles['content-wrapper'],
+          media?.content && styles['content-wrapper-with-media'],
+          fitHeight && styles['content-wrapper-fit-height']
+        )}
       >
-        {children}
-      </div>
-      {footer && (
+        {header && (
+          <StickyHeaderContext.Provider value={{ isStuck }}>
```
Comment (timogasda) [inline src/container/internal.tsx:137]: So what happens if a sticky header has media in it? 🤔

File: src/container/media.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import { useEffect, useState } from 'react';
+import { Breakpoint, BREAKPOINTS_DESCENDING, matchBreakpointMapping } from '../internal/breakpoints';
+import { ContainerProps, MediaDefinition } from './interfaces';
+import { useContainerBreakpoints } from '../internal/hooks/container-queries';
+
+type MediaProps = keyof ContainerProps.Media;
+
+const MEDIA_PROPS: MediaProps[] = ['position', 'width', 'height', 'content'];
+
+const hasDefinedBreakpoint = (obj: any) => {
```
Comment (timogasda) [inline src/container/media.ts:13]: I think we can do better than `any`. Maybe something like `obj: { breakpoint: Breakpoint }` ? Or a generic type

File: src/container/media.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    position: media?.position || 'top',
+  };
+  const [mediaState, setMediaState] = useState(initialState);
+  const [breakpoint, breakpointRef] = useContainerBreakpoints(getBreakpointsForMedia(media));
+  useEffect(() => {
+    if (!media || !breakpoint) {
+      return;
+    }
+
+    setMediaState(prevMediaState => {
```
Comment (timogasda) [inline src/container/media.ts:31]: Please consider if you really need a `useEffect` hook here: https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes.

File: src/container/media.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+      MEDIA_PROPS.forEach(attribute => {
+        if (!media[attribute]) {
+          return;
+        }
+
+        let value;
+
+        if (attribute === 'content') {
+          value = hasDefinedBreakpoint(media[attribute])
+            ? matchBreakpointMapping(media[attribute] as MediaDefinition.BreakpointMapping<any>, breakpoint)
```
Comment (timogasda) [inline src/container/media.ts:43]: `as` (and `any`) are stripping away a lot of type safety here. Can we get rid of them?

File: src/container/media.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+      const newMediaState = { ...prevMediaState };
+
+      MEDIA_PROPS.forEach(attribute => {
+        if (!media[attribute]) {
+          return;
+        }
+
+        let value;
+
+        if (attribute === 'content') {
```
Comment (timogasda) [inline src/container/media.ts:41]: This `if/else` is the meat of this hook. Can we please add some descriptive comments?

File: src/container/media.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    return;
+  }
+
+  const breakpointKeys = new Set<Breakpoint>();
+
+  MEDIA_PROPS.forEach(key => {
+    if (media && typeof media[key] === 'object' && media[key] !== null) {
+      const breakpointMapping = media[key] as MediaDefinition.BreakpointMapping<any>;
+      Object.keys(breakpointMapping).forEach(key => {
+        const breakpoint = key as Breakpoint;
```
Comment (timogasda) [inline src/container/media.ts:79]: Two more uses of `as` in this function. This might be an indicator that your props and/or arguments are not typed correctly.

File: src/container/styles.scss [hunk 9]
```diff
@@ -61,11 +64,68 @@
...
+  &-with-media {
+    display: flex;
+    flex-direction: column;
+  }
+}
+
+.media {
+  overflow: hidden;
+
+  // stylelint-disable-next-line @cloudscape-design/no-implicit-descendant, selector-max-type
```
Comment (timogasda) [inline src/container/styles.scss:91]: Shouldn't we be using a direct descending selector (`>`) here?

Side note: Should we add some additional restrictions to the API docs or the Dev guidelines? Things like: "we only support `img`, `video`, `picture`", or "the media tag needs to be the first and only child of the `media` slot"

File: src/container/styles.scss [hunk 10]
```diff
@@ -61,11 +64,68 @@
...
   border-top-left-radius: awsui.$border-radius-container;
   border-top-right-radius: awsui.$border-radius-container;
 
+  &.header-with-media {
+    background: none;
+    &:not(:empty) {
+      border-bottom: none;
+    }
+    &.content {
+      overflow: auto;
```
Comment (timogasda) [inline src/container/styles.scss:125]: Why is this needed? Why do we want containers with media to scroll their content?

File: src/expandable-section/__tests__/expandable-section.test.tsx [hunk 11]
```diff
@@ -29,7 +33,7 @@ describe('Expandable Section', () => {
       expect(wrapper.findAll('button').length).toBe(0);
     });
   });
-  describe('slots', () => {
+  describe.only('slots', () => {
```
Comment (timogasda) [inline src/expandable-section/__tests__/expandable-section.test.tsx:36]: `.only` ! 

(we really should add an eslint rule about this...)

File: src/expandable-section/__tests__/expandable-section.test.tsx
Comment (timogasda) [inline src/expandable-section/__tests__/expandable-section.test.tsx:1]: Should we add some new tests?

**Review by timogasda — CHANGES_REQUESTED** [review]

File: src/container/internal.tsx (hunk 3 above)
Comment (rubencarvalho) [reply src/container/internal.tsx:137]: This combination should not be possible for consumers of the Container.

`__stickyHeader` is a private property that we only use internally for having the sticky container throughout other components (like Cards or Table). I don't see us wanting both to coexist in these scenarios, so I think it is safe to ignore the corner case.

File: src/container/internal.tsx (hunk 1 above)
Comment (rubencarvalho) [reply src/container/internal.tsx:107]: Leftover from refactoring, will remove this!

File: src/container/styles.scss (hunk 9 above)
Comment (rubencarvalho) [reply src/container/styles.scss:91]: I've thought about this. My thought was that there are some custom media components (like those offered by some frameworks, i.e.: `GatsbyImage` or `next/image`) that could potentially include more wrapping elements.
What do you think?

Regarding documentation, I will update! 👍

Comment (timogasda) [reply src/container/styles.scss:91]: mhm yeah, fair enough. this might lead to interesting questions down the line in terms of what we support. in case some people have weird mixed media components.

But for now I think this is fine.

File: src/container/internal.tsx (hunk 2 above)
Comment (rubencarvalho) [reply src/container/internal.tsx]: Nice, I updated it. And just like that... way cleaner!

Thanks!

File: src/container/media.ts (hunk 5 above)
Comment (rubencarvalho) [reply src/container/media.ts]: Indeed! Great call - no need for state even.

File: src/container/interfaces.ts [hunk 12]
```diff
@@ -47,3 +50,48 @@ export interface ContainerProps extends BaseComponentProps {
    */
   variant?: 'default' | 'stacked';
 }
+
+export namespace ContainerProps {
+  /**
+   * The value for each property can either be a single value (which applies for all breakpoints) or an object where the key
+   * is one of the supported breakpoints (`xxs`, `xs`, `s`, `m`, `l`, `xl`) or `default`.
+   * The value of this key is applied for that breakpoint and those above it. You must provide a `default` value for `content`.
+   */
```
[the remark this answers is not in the document]
Comment (rubencarvalho) [reply src/container/interfaces.ts]: Good call. I'd rather show it. I've moved it down to the `content` slot.

File: src/container/interfaces.ts [hunk 13]
```diff
@@ -47,3 +50,48 @@ export interface ContainerProps extends BaseComponentProps {
...
+
+    /**
+     * Defines the media slot's orientation within the container.
+     */
+    orientation: MediaDefinition.Orientation | MediaDefinition.BreakpointMapping<MediaDefinition.Orientation>;
+
+    /**
+     * Defines the width of the media slot when vertically oriented. Corresponds to the `width` CSS-property.
+     * When this value is set, the media element may be cropped, with `object-fit: cover` centering it.
+     * Note: This value is considered only when `orientation` is set to `vertical`.
```
[the remark this answers is not in the document]
Comment (rubencarvalho) [reply src/container/interfaces.ts]: Done!

File: src/container/internal.tsx (hunk 1 above)
Comment (rubencarvalho) [reply src/container/internal.tsx]: Done!

File: src/container/styles.scss (hunk 10 above)
Comment (rubencarvalho) [reply src/container/styles.scss]: Removed!

File: src/container/media.ts (hunk 8 above)
Comment (timogasda) [reply src/container/media.ts]: Nevermind. `Object.keys` is just really useless in Typescript 😬

File: src/container/internal.tsx [hunk 14]
```diff
@@ -74,7 +78,10 @@ export default function InternalContainer({
   const hasDynamicHeight = isRefresh && variant === 'full-page';
   const overlapElement = useDynamicOverlap({ disabled: !hasDynamicHeight || !__darkHeader });
 
-  const mergedRef = useMergeRefs(rootRef, subStepRef, __internalRootRef);
+  const [breakpoint, breakpointRef] = useContainerBreakpoints(BREAKPOINTS_DESCENDING.filter(bp => bp !== 'default'));
```
Comment (timogasda) [inline src/container/internal.tsx:81]: Would it make sense to move the `useContainerBreakpoints` into `useMedia`?

Comment (timogasda) [reply src/container/internal.tsx:81]: also, instead of listening to all breakpoints (except default), could you extract the list of breakpoints use from the `media` prop and use them instead?

**Review by timogasda — COMMENTED** [review]
Looks great overall. Just one small suggestion

File: src/container/interfaces.ts [hunk 15]
```diff
@@ -47,3 +74,47 @@ export interface ContainerProps extends BaseComponentProps {
...
+    content: React.ReactNode | MediaDefinition.ContentBreakpointMapping;
+
+    /**
+     * Defines the media slot's position within the container. Defaults to `top`.
+     */
+    position?: 'top' | 'side' | MediaDefinition.BreakpointMapping<'top' | 'side'>;
+
+    /**
+     * Defines the width of the media slot when positioned on the side. Corresponds to the `width` CSS-property.
+     * When this value is set, the media element may be cropped, with `object-fit: cover` centering it.
```
Comment (stirache) [inline src/container/interfaces.ts:97]: When this value is set, media elements larger than the defined width may be cropped, with 'object-fit: cover' centering it.

**Review by stirache — CHANGES_REQUESTED** [review]

**Review by timogasda — APPROVED** [review]
Looks good!
